libs/libc/pthread/pthread_once.c

```diff
@@ -73,15 +75,15 @@ int pthread_once(FAR pthread_once_t *once_control,
 
   /* Prohibit pre-emption while we test and set the once_control. */
 
-  sched_lock();
+  flags = enter_critical_section();
```

Comment:
The only one question here: is critical section enough or we need to use spinlock? I mean on theory can we have a situation when `once_control` may be modified from different cores?

Reply:
Yes, it's possible with SMP.

Reply:
So should we use spinlock here?

Reply:
but it's dangerous to call user provided callback inside spinlock.

Reply:
But it is only atomic variable check and modification protected by a spinlock and not a user routine

Reply:
The implementation is wrong, pthread_once should block the second thread until the first second finish init_rountine invocation, otherwise the second thread may access the partial initialized struct or object.

Reply:
So neither scheduler lock nor critical section nor spinlock will work with current implementation.

Reply:
yes it needs a full mutex implementation, so context aware.

Reply:
@hujun260 please provide a more correct fix.